File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/WithdrawalLimiter.java
package ro.razvan.java.dagger.atm;

import javax.inject.Inject;
import java.math.BigDecimal;

@PerSession
public final class WithdrawalLimiter {

    private BigDecimal remainingWithdrawalLimit;

    @Inject
    public WithdrawalLimiter(@MaximumWithdrawal BigDecimal maximumWithdrawal) {
        remainingWithdrawalLimit = maximumWithdrawal;
    }

    public void recordDeposit(final BigDecimal amount) {
        remainingWithdrawalLimit = remainingWithdrawalLimit.add(amount);
    }

    public void recordWithdrawal(final BigDecimal amount) {
        remainingWithdrawalLimit = remainingWithdrawalLimit.subtract(amount);
    }

    public BigDecimal remainingWithdrawalLimit() {
        return remainingWithdrawalLimit;
    }
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/WithdrawalLimiter.kt
package ro.razvan.kotlin.dagger.atm

import java.math.BigDecimal
import javax.inject.Inject

@PerSession
class WithdrawalLimiter @Inject constructor(
    @MaximumWithdrawal maximumWithdrawal: BigDecimal
) {

    var remainingWithdrawLimit: BigDecimal
        private set

    init {
        remainingWithdrawLimit = maximumWithdrawal
    }

    fun recordDeposit(amount: BigDecimal) {
        remainingWithdrawLimit += amount
    }

    fun recordWithdrawal(amount: BigDecimal) {
        remainingWithdrawLimit -= amount
    }

}